### DISCUSSION TITLE
Troubleshoot Flashing Jetson TX2 Module?

### DISCUSSION THREAD

**Discussion by tmdarwen** [OE4T/meta-tegra#876] [Q&A] [answered]
Using yocto, I've successfully built an image for a Jetson TX2 Module.  I'm trying to flash the image using doflash.sh but running into the "_USB communication failed. Check if device is in recovery_" error.

I'm quite certain the device is in recovery mode as I see the following when running lsusb:

```
buildmachine@buildmachine:~/NonSys/Jetson/build$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 002: ID 413c:2514 Dell Computer Corp. Dell Universal Receiver
Bus 001 Device 005: ID 0955:7020 NVIDIA Corp. L4T (Linux for Tegra) running on Tegra
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
When running doflash.sh I get the following:
```
Welcome to Tegra Flash
version 1.0.0
Type ? or help for help and q or quit to exit
Use ! to execute system commands

[   0.0024 ] Generating RCM messages
[   0.0031 ] tegrarcm_v2 --listrcm rcm_list.xml --chip 0x18 0 --download rcm mb1_recovery_prod.bin 0 0
[   0.0037 ] RCM 0 is saved as rcm_0.rcm
[   0.0039 ] RCM 1 is saved as rcm_1.rcm
[   0.0039 ] List of rcm files are saved in rcm_list.xml
[   0.0039 ]
[   0.0040 ] Signing RCM messages
[   0.0046 ] tegrasign_v2 --key None --list rcm_list.xml --pubkeyhash pub_key.key
[   0.0051 ] Assuming zero filled SBK key
[   0.0078 ]
[   0.0079 ] Copying signature to RCM mesages
[   0.0085 ] tegrarcm_v2 --chip 0x18 0 --updatesig rcm_list_signed.xml
[   0.0093 ]
[   0.0093 ] Boot Rom communication
[   0.0099 ] tegrarcm_v2 --chip 0x18 0 --rcm rcm_list_signed.xml
[   0.0105 ] BootRom is not running
[   0.0108 ]
[   1.0151 ] tegrarcm_v2 --isapplet
[   1.0175 ] USB communication failed.Check if device is in recovery
[   1.0182 ]
[   1.0208 ] tegradevflash_v2 --iscpubl
[   1.0231 ] Cannot Open USB
```

My build machine from which I'm running doflash.sh from is Ubuntu 20.04.  I've tried two different USB cables with the same result.  I'm not finding much help when Googling.

Any suggestions on how to troubleshoot this issue?

**Comment by kekiefer** [comment]
It's probable that your login user does not have permissions it needs to access the USB device. The easiest thing you can do is run the flash script with `sudo`. Another option would be to add some udev rules to give your login user permissions to access the device. I have the following rule installed as `/etc/udev/rules.d/40-nvidia.rules` (my login user is a member of the `plugdev` group):

```
SUBSYSTEM=="usb", ATTR{idVendor}=="0955", MODE="0660", GROUP="plugdev"
```

**Comment by tmdarwen** [comment]
Hi, thanks for the reply, apologies for the slow-ish follow-up response as I submitted my question on a Friday evening just before leaving the office.  Yes, I did run doflash.sh as sudo.  I can't imagine using the udev rule would allow some permission that sudo does not have.  Right?

Also, any other suggestions would be quite welcome, thank you.

**Comment by madisongh** [comment]
Based on the output from `lsusb`, it looks like your unit isn't actually in recovery mode:
```
Bus 001 Device 005: ID 0955:7020 NVIDIA Corp. L4T (Linux for Tegra) running on Tegra
```
That USB identifier is used for the USB device mode setup when Linux is running.  Either log in and do `sudo reboot forced-recovery` to have it reboot into recovery mode, or use the reset and recovery buttons as mentioned in the L4T docs and on [the wiki](https://github.com/OE4T/meta-tegra/wiki/Flashing-the-Jetson-Dev-Kit#jetson-tx1-and-tx2-development-kits).  The ID in `lsusb` output should be `0955:7c18` for a TX2 when it's in recovery.

**Comment by tmdarwen** [comment, marked as answer]
Thanks @madisongh - that was the issue and I've now been able  to run doflash.sh.  However, after "sending blob" it just hangs with this final line: "tegrarcm_v2 --isapply" as shown below.  Is this normal?  Any advice?

```
[   6.6702 ] tegrahost_v2 --chip 0x18 --generateblob blob.xml blob.bin
[   6.6707 ] number of images in blob are 9
[   6.6708 ] blobsize is 4401240
[   6.6709 ] Added binary blob_nvtboot_recovery_cpu_sigheader.bin.encrypt of size 222432
[   6.6722 ] Added binary blob_nvtboot_recovery_sigheader.bin.encrypt of size 122256
[   6.6724 ] Added binary blob_preboot_d15_prod_cr_sigheader.bin.encrypt of size 63104
[   6.6726 ] Added binary blob_mce_mts_d15_prod_cr_sigheader.bin.encrypt of size 2082144
[   6.6730 ] Added binary blob_bpmp_sigheader.bin.encrypt of size 534816
[   6.6733 ] Added binary blob_tegra186-a02-bpmp-quill-p3310-1000-a00-00-te770d-ucm2_sigheader.dtb.encrypt of size 604880
[   6.6736 ] Added binary blob_tos-trusty_sigheader.img.encrypt of size 395072
[   6.6738 ] Added binary blob_eks_sigheader.img.encrypt of size 1440
[   6.6740 ] Added binary blob_tegra186-quill-p3310-1000-c03-00-base_sigheader.dtb.encrypt of size 374944
[   6.6753 ]
[   6.6753 ] Sending bootloader and pre-requisite binaries
[   6.6759 ] tegrarcm_v2 --download blob blob.bin
[   6.6763 ] Applet version 01.00.0000
[   6.6941 ] Sending blob
[   6.6941 ] [................................................] 100%
[   7.2132 ]
[   7.2160 ] tegrarcm_v2 --boot recovery
[   7.2182 ] Applet version 01.00.0000
[   7.2357 ]
[   8.2399 ] tegrarcm_v2 --isapplet
```

**Reply by madisongh** [reply]
Do you have access to the serial console?  It would help to see what's being logged there.

**Reply by tmdarwen** [reply]
Hi, thanks @madisongh actually I kept working on it last night and think I got a little further.  After I ran into the doflash issue mentioned in my last post, as a sanity check I followed [these NVIDIA instructions](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/quick_start.html#wwpID0E0BE0HA) flashing the NVIDIA Jetson Linux onto the board successfully.

During this process I realized I likely built my Yocto image using the wrong machine name.  I thought we had a TX2 module and I used "jetson-tx2-devkit".  In reality we have a "TX2-NX module" ([Digi-key link](https://www.digikey.com/en/products/detail/seeed-technology-co-ltd/102110547/13926229?s=N4IgTCBcDaIIwAYx0QVgCwHYQF0C%2BQA)).  We're using the TX2-NX module in a [Jetson Mate](https://www.seeedstudio.com/Jetson-Mate-Cooling-Kit-p-4784.html?gclid=CjwKCAjwq9mLBhB2EiwAuYdMtYdpfjhreKK-YnmwL7T7jmvAFQxPo_JI9WkA89uMOJNmXOAUXQCObhoCf3YQAvD_BwE).

So, reviewing the possible machine types for the Yocto build as mentioned [here](https://developer.ridgerun.com/wiki/index.php?title=Yocto_Support_for_NVIDIA_Jetson_Platforms_-_Setting_up_Yocto#Select_your_Machine), it occurred to me I need to be using an "eMMC" machine name.  Following the instruction [here](https://developer.ridgerun.com/wiki/index.php?title=Yocto_Support_for_NVIDIA_Jetson_Platforms_-_Setting_up_Yocto#Select_your_Machine), the only eMMC packages are "jetson-nano-devkit-emmc" and "jetson-xavier-nx-devkit-emmc".  We're certainly not using a Nano, so I did a Yocto build using the machine name "jetson-xavier-nx-devkit-emmc".   However, when attempting to flash it, I end up getting this error: `ERR: chip ID mismatch for Xavier`

So, it sounds like the exact machine name we need does not yet exist in the meta-tegra repo.  I imagine we need some sort of `jetson-tx2-nx-devkit-emmc.conf` file in `https://github.com/OE4T/meta-tegra/tree/master/conf/machine`.  Do you agree?  And if so, is this something you can help guide me in creating?  I'd be happy to add it to the meta-tegra repo with a pull request once it's working.

**Reply by madisongh** [reply]
The machine name you're looking for is [jetson-xavier-nx-devkit-tx2-nx](https://github.com/OE4T/meta-tegra/blob/master/conf/machine/jetson-xavier-nx-devkit-tx2-nx.conf); that's the one for the TX2-NX installed in a Xavier NX devkit carrier.  Hopefully the Jetson Mate is similar enough that the existing boot configuration and device tree files will be compatible.  (Edited to correct the name; link was correct)

**Reply by tmdarwen** [reply]
Worked!  I've successfully flashed the device with the resulting Yocto Linux image.

Thank you @madisongh - really appreciate the assistance.
